### PR TITLE
API chapter

### PR DESCRIPTION
**Pull request by nevenRakonic**

### CODE REVIEW

File: Building an API.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+Building an API is different than building a regular HTML application. There are
+no HTML views to worry about, but instead you should take extra care about status responses, response format and especially documentation since your application will be used
+by other developers.
+
+## Creating an application
+
+To generate a Rails application with only the elements necessary for an API
+application use the `--api` flag e.g.
+`rails new my_application --api`
```
Comment (d4be4st) [inline Building an API.md]: This may be true, but you will need an admin panel for 99% of applications, so the --api flag wont do

Comment (DamirSvrtan) [inline Building an API.md]: I'm with @d4be4st on this, and I would put this rationale in this document.

File: Building an API.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+You can generate documentation from your controller or request specs, and it's very
+important that you cover all the possible cases and return codes e.g. a valid response,
+response when there are errors and any other outcome of the action. Also document any filtering parameters and enumerations that may be used to refine the results of the API call.
+
+https://apiary.io/ is a service that can be used to host the API documentation.
+
+## Versioning
+
+APIs should be versioned from the start, that way in case of large changes clients can still use the older version of the API while adjusting their application to the new API specification. Otherwise if the API is not versioned, changes to the API would break clients applications frequently and the client would have to take a lot of care in tracking the changes.
```
Comment (melcha) [inline Building an API.md]: *must be versioned

File: Building an API.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+You can make different API versions obvious by namespacing them in the URL e.g. ```api/V1/resource``` where the new version would look like ```api/V2/resource```
```
Comment (melcha) [inline Building an API.md]: We need a firmer statement here, like: 
> Namespace different API versions in the URL, e.g. ```api/v1/resource``` 

(lower case v1). And make sure your controllers, serializers, documentation and tests are also namespaced in the application code.

File: Building an API.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+## Pagination
+
+It's often possible that the results of the API call are huge e.g. hundreds of thousands of records, and that's impractical if the client only needs a small part of the results. That's why we can limit the amount of results returned with pagination, which allows the client to choose which part of the results they want to see in the request e.g results 500-600 out of 10000.
```
Comment (melcha) [inline Building an API.md]: Pagination is a *must have* for (almost) all index actions. Pagination should be described in the API documentation (which query params are being used), list default page size and max page size.

File: Building an API.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+## HTTP status codes and headers
```
Comment (melcha) [inline Building an API.md]: Json API standard actually specifies `application/vnd.api+json` for Content-Type/Accept headers. API should have Content-Type set with this value in all responses.

We should also have a list of status codes we usually use in our APIs:
- 200, 201 (created), 204  (no content, on destroy actions)
- 422 (resource is not saved), 401 (user is not logged in), 403 (user doesn't have a permission for the action), 400 (bad request, unsupported sort param), 404

File: Building an API.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+* Testing is always important when building software, but doubly so when writing
+an API since we can use tests to generate documentation.
+* [dox](https://github.com/infinum/dox)
+* Generates documentation from rspec tests in api blueprint format.
+
+## Documentation
+
+You can generate documentation from your controller or request specs, and it's very
+important that you cover all the possible cases and return codes e.g. a valid response,
+response when there are errors and any other outcome of the action. Also document any filtering parameters and enumerations that may be used to refine the results of the API call.
```
Comment (melcha) [inline Building an API.md]: Make a list of what should be in the documentation, so it's more noticeable. 

Add also supported sorting parameters (as requesting sort with unsupported param should return 400), how the authentication is done, pagination parameters (with default and max page size). 

Re hosting documentation - it's best to put the documentation regeneration and push in the  CI process (after successful build), so the documentation is always up to date.

File: Building an API.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+no HTML views to worry about, but instead you should take extra care about status responses, response format and especially documentation since your application will be used
+by other developers.
+
+## Creating an application
+
+To generate a Rails application with only the elements necessary for an API
+application use the `--api` flag e.g.
+`rails new my_application --api`
+
+We write our APIs in JSON format, more specifically using the [JSON API standard](http://jsonapi.org/).
```
Comment (DamirSvrtan) [inline Building an API.md]: We write our APIs in JSON format -> We write our APIs using the JSON format.

File: Building an API.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+This allows those who consume our APIs to know what to expect regarding the
```
Comment (DamirSvrtan) [inline Building an API.md]: I love the fact that you added reasoning to why we use the JSON API standard. 
Please also note that the reason we use the standard is that the devs building the client apps can use libraries that are built to consume the JSON API standard, making it easier and faster to build their clients.

File: Building an API.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+document structure.
+
+## Gems
+
+We use a few gems that make our lives easier when writing APIs:
+* [json_api_responders](https://github.com/infinum/json_api_responders)
+* Simplifies the code in controllers, no need to write `if @object.save` etc. to
```
Comment (DamirSvrtan) [inline Building an API.md]: is replaced with `respond_with resource`.

**Comment by DamirSvrtan** [comment]
@nevenRakonic what's the status with this?

**Comment by nevenRakonic** [comment]
@DamirSvrtan I didn't have much time to work on this lately, I'll update it with the comments in mind.

**Comment by melcha** [comment]
👍 

@janvarljen please check it out. Do we have everything important covered?

**Comment by janvarljen** [comment]
@melcha 

Depends how deep you want to go into details. For starters, this looks fine to me

**Pull request merged**